### DISCUSSION TITLE
Le parseur Nearley n'a pas réussi à parser l'expression

### DISCUSSION THREAD

**Discussion by Badio75** [publicodes/publicodes#445] [Q&A] [answered]
Bonjour et félicitation pour ce projet très prometteur ! 

J'essaye de l'utiliser dans une API Express.js en suivant la documentation. Cela me donne le code suivant : 
```
// const Engine = require('publicodes').default

const rules = await import('modele-social');
const engine = new Engine(rules)

engine.setSituation({
    ...situation
})

const evaluation = engine.evaluate(expressions)
const evaluate = evaluation.evaluate
```

J'obtient toutefois l'erreur suivante : 

> le parseur Nearley n'a pas réussi à parser l'expression.
> 
>     at parseExpression (/x/x/x/x/x/x/node_modules/publicodes/dist/index.cjs:4423:13)

Arrivez vous à voir ce que je fais mal ?

**Comment by johangirod** [comment]
Bonjour :)

Pour répondre, il faudrait connaître le contenue de `situation`, c'est sans doute à ce niveau que se situe l'erreur.

Par ailleurs, pour accéder à la valeur évaluer il faut utiliser `evaluation.nodeValue` et non `evaluation.evaluate`

**Comment by Badio75** [comment]
Merci pour votre réponse. Ci dessous le contenue de `situation` et de `Expressions`. 

situation = {
  'dirigeant . régime social': "'indépendant'",
  'entreprise . activité . nature': "'commerciale'",
  'entreprise . imposition': "'IS'",
  'entreprise . catégorie juridique': "''",
  'entreprise . date de création': '01/01/2024',
  'dirigeant . exonérations . ACRE': 'non',
  'dirigeant . rémunération . totale': 150000
} 

expressions = [
  'dirigeant . indépendant . assiette des cotisations',
  'dirigeant . indépendant . cotisations et contributions . début activité',
  'dirigeant . indépendant . cotisations et contributions',
  'dirigeant . indépendant . cotisations et contributions . maladie',
  'dirigeant . indépendant . cotisations et contributions . retraite de base',
  'dirigeant . indépendant . cotisations et contributions . retraite complémentaire',
  'dirigeant . indépendant . cotisations et contributions . indemnités journalières maladie',
  'dirigeant . indépendant . cotisations et contributions . invalidité et décès',
  'dirigeant . indépendant . cotisations et contributions . allocations familiales',
  'dirigeant . indépendant . cotisations et contributions . formation professionnelle',
  'dirigeant . indépendant . cotisations et contributions . CSG-CRDS',
  'dirigeant . rémunération . net'
]

**Comment by johangirod** [comment]
`evaluate` prend comme paramètre une règle, et non un tableau de règle. Pour évaluer plusieurs règles, il faut appeler `evaluate` pour chacune d'entre elle : 

```js

engine.setSituation({
    ...situation
})

const evaluations = expressions.map(expression => engine.evaluate(expression).nodeValue)

```

**Reply by Badio75** [reply]
Merci beaucoup ! En effet, j'utilisais avant l'API mon-entreprise pour des tests et celle-ci acceptait un tableau de règles. J'ai modifié mon code mais je conserve la même erreur. Pour plus de clarté, mon code est maintenant le suivant : 

```
const Engine = require('publicodes').default

exports.computeResult = async () => {

    const rules = await import('modele-social');
    const engine = new Engine(rules)
    
    engine.setSituation({
        "dirigeant . indépendant . IJSS": "non",
        "dirigeant . indépendant . revenus étrangers": "non",
        "entreprise . imposition": "'IS'",
        "situation personnelle . domiciliation fiscale à l'étranger": "non",
        "entreprise . activité . nature": "'commerciale'",
        "entreprise . catégorie juridique": "''",
        "entreprise . date de création": "01/01/2024",
        "dirigeant . exonérations . ACRE": "non",
        "dirigeant . régime social": "'indépendant'",
        "dirigeant . rémunération . totale": 150000,
    })


    const evaluation = engine.evaluate("dirigeant . indépendant . cotisations et contributions")
    console.log(evaluation)
}
```

```
le parseur Nearley n'a pas réussi à parser l'expression.
...
info: { dottedName: 'json' }
```

**Reply by johangirod** [reply, marked as answer]
Il s'agit d'un problème avec votre utilisation de JavaScript, et non de la publicodes. Il faut ajouter `.default` à rules. 
```js
    const rules = (await import('modele-social')).default;
```
